File: leetcode/0635-design-log-storage-system/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class LogSystem {
    private static final Map<String, Integer> GRANULARITY = new HashMap<>();
    static {
        GRANULARITY.put("Year", 0);
        GRANULARITY.put("Month", 1);
        GRANULARITY.put("Day", 2);
        GRANULARITY.put("Hour", 3);
        GRANULARITY.put("Minute", 4);
        GRANULARITY.put("Second", 5);
    }

    private final List<Integer> ids = new ArrayList<>();
    private final List<int[]> parts = new ArrayList<>();

    public LogSystem() {
    }

    private int[] parse(String timestamp) {
        String[] tokens = timestamp.split(":");
        int[] result = new int[6];
        for (int i = 0; i < 6; i++) result[i] = Integer.parseInt(tokens[i]);
        return result;
    }

    // lexicographic comparison of the first (idx + 1) fields, like comparing tuples
    private int compare(int[] a, int[] b, int idx) {
        for (int i = 0; i <= idx; i++) {
            if (a[i] != b[i]) return Integer.compare(a[i], b[i]);
        }
        return 0;
    }

    public void put(int id, String timestamp) {
        ids.add(id);
        parts.add(parse(timestamp));
    }

    public List<Integer> retrieve(String start, String end, String granularity) {
        int idx = GRANULARITY.get(granularity);
        int[] startParts = parse(start);
        int[] endParts = parse(end);
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < parts.size(); i++) {
            int[] p = parts.get(i);
            if (compare(p, startParts, idx) >= 0 && compare(p, endParts, idx) <= 0) {
                result.add(ids.get(i));
            }
        }
        return result;
    }
}

/**
 * Your LogSystem object will be instantiated and called as such:
 * LogSystem obj = new LogSystem();
 * obj.put(id,timestamp);
 * List<Integer> param_2 = obj.retrieve(start,end,granularity);
 */

File: leetcode/0635-design-log-storage-system/solution.cpp
#include <array>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class LogSystem {
    vector<pair<int, array<int, 6>>> logs; // (id, [year, month, day, hour, minute, second])
    unordered_map<string, int> granularity = {
        {"Year", 0}, {"Month", 1}, {"Day", 2}, {"Hour", 3}, {"Minute", 4}, {"Second", 5}
    };

    array<int, 6> parse(const string& timestamp) {
        array<int, 6> result{};
        int idx = 0, i = 0;
        while (i < (int)timestamp.size()) {
            int j = i;
            while (j < (int)timestamp.size() && timestamp[j] != ':') j++;
            result[idx++] = stoi(timestamp.substr(i, j - i));
            i = j + 1;
        }
        return result;
    }

    // lexicographic comparison of the first (idx + 1) fields, like comparing tuples
    int compare(const array<int, 6>& a, const array<int, 6>& b, int idx) {
        for (int i = 0; i <= idx; i++) {
            if (a[i] != b[i]) return a[i] < b[i] ? -1 : 1;
        }
        return 0;
    }

public:
    LogSystem() {}

    void put(int id, string timestamp) {
        logs.push_back({id, parse(timestamp)});
    }

    vector<int> retrieve(string start, string end, string gran) {
        int idx = granularity[gran];
        array<int, 6> startParts = parse(start);
        array<int, 6> endParts = parse(end);
        vector<int> result;
        for (auto& [id, parts] : logs) {
            if (compare(parts, startParts, idx) >= 0 && compare(parts, endParts, idx) <= 0) {
                result.push_back(id);
            }
        }
        return result;
    }
};

/**
 * Your LogSystem object will be instantiated and called as such:
 * LogSystem* obj = new LogSystem();
 * obj->put(id,timestamp);
 * vector<int> param_2 = obj->retrieve(start,end,granularity);
 */
